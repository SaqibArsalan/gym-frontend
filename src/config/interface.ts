import { Theme } from '@mui/material';
import { ROLES } from './enums';

export interface IChildModuleInfo {
	key: string;
	moduleKey: string;
	routeUrl: string;
	exposedMainComponent: string;
	allowedRoles: ROLES[];
}

export interface IAppProps {
	mainTheme?: Theme;
}
