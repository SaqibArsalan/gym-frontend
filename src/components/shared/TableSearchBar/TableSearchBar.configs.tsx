import React, { useState } from 'react';
import SearchIcon from '@mui/icons-material/Search';
import ClearIcon from '@mui/icons-material/Clear';
import { TextFieldProps, Grid, InputAdornment, IconButton, MenuItem, FormControl } from "@mui/material";
import { StyledSelectField } from './TableSearchBar.styles';
import { IFilterDropdown, IFilterOption } from './TableSearchBar.interface';

export const FilterDropdown = (props: IFilterDropdown) => {
	const { defaultValue, onFilterChange, options } = props;
	const [value, setValue] = useState<string>(defaultValue ? defaultValue.value : "");
	const handleChange = (event: any) => {
		const changedValue = event.target.value as string;
		setValue(changedValue);
		if(onFilterChange) {
			const option = options.find((item: IFilterOption) => item.value === changedValue);
			if(option) {
				onFilterChange(option);
			}
		}
	  };
	return (
		<FormControl fullWidth>
			<StyledSelectField
				value={value}
				onChange={handleChange}
			>
				{options ? options.map((item: any, index: number) => <MenuItem key={index} value={item.value}>{item.name}</MenuItem>): null}
			</StyledSelectField>
		</FormControl>
	);
}


export const applyFilterHOC = (TextField: React.FunctionComponent<TextFieldProps>) => 
	(props: {textFieldProps: TextFieldProps, filterProps: IFilterDropdown}) => {
	const {textFieldProps, filterProps} = props;
	return (
		<Grid
			container
			className='table-search-bar-container'
			columns={12}
			spacing={0}
		>
				<Grid item sm={9} md={9} lg={10} xs={12}>
					<TextField
						{...textFieldProps}
					/>
				</Grid>
				<Grid className='filter-dropdown' item sm={3} lg={2} md={3} xs={12}>
					<FilterDropdown {...filterProps} />
				</Grid>
		</Grid>
	);
}

export const InputAdornmentConfiguration = (
	showRemove: boolean, 
	onRemoveIconClick: React.MouseEventHandler<HTMLAnchorElement>, 
	removeIcon?: any, 
	searchIcon?: any
) => {
	const RemoveIconElement: any = removeIcon || ClearIcon;
	const SearchIconElement: any = searchIcon || SearchIcon;
return {
	startAdornment: (
		<InputAdornment position='start'>
			<SearchIconElement />
		</InputAdornment>
	),
	endAdornment:
		showRemove ? (
			<>
				<InputAdornment position='end'>
					<IconButton
						id='remove-button'
						aria-label='remove value'
						component='span'
						onClick={onRemoveIconClick}
					>
						<RemoveIconElement />
					</IconButton>
				</InputAdornment>
			</>
		) : null,
}};