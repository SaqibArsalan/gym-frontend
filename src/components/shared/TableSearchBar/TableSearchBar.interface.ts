// Types

import React from 'react';

/* eslint-disable-next-line */
type onChange = (value: string, filterOption?: any) => void;
/* eslint-disable-next-line */
type PlaceholderFunction = () => void;

// Table Search Bar interface

export interface ITableSearchBarProps {
	enableFilter?: boolean;
	filterOptions?: IFilterOption[];
	defaultSelectedFilterOption?: IFilterOption;
	onFilterChange?: (item: any) => void;
	defaultValue?: string;
	placeholder?: string;
	onChange?: onChange;
	searchIcon?: any;
	showAddButton?: boolean;
	onAddButtonClick?: PlaceholderFunction;
	removeIcon?: any;
	showRemoveIcon?: boolean;
	onRemove?: PlaceholderFunction;
	customInputComponent?: React.FC<any>;
	customInputComponentProps?: any;
}


export interface IFilterDropdown {
	options: IFilterOption[]; defaultValue: any; onFilterChange: (option: IFilterOption) => void;
}

export interface IFilterOption {
	name: string;
	value: string;
}