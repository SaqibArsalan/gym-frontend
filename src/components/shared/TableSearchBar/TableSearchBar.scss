@import 'src/assets/scss/Variables.scss';
@import 'src/assets/scss/Extends.scss';

.toolbar-content-wrapper {
	width: 100%;
	display: flex;
	.textfield-wrapper {
		flex-basis: 100%;
		.table-search-bar-container {
			&:hover {
				border: 1px solid black;
			}
			border: 1px solid lightgray;
			border-radius: 4px;
			fieldset {
				border: 0px !important;
			}
			.filter-dropdown {
				.MuiSelect-select {
					border-radius: 0px;
					border-left: 1px solid lightgray;
				}
			}
		}
	}
	.add-button-wrapper {
		flex-basis: 0%;
		position: relative;
		margin-top: 10px;
		button {
			right: 0px;
			top: 38px;
			position: absolute;
			z-index: 99;
			background-color: $skyblue;
		}
	}
}

@media (max-width: $mobile-view) {
	.toolbar-content-wrapper {
		.textfield-wrapper {
			.table-search-bar-container {
				.filter-dropdown {
					input {
						border: 0px;
					}
					.MuiSelect-select {
						border-top: 1px solid lightgray;
						border-radius: 0px;
					}
				}
			}
		}
	}
}
