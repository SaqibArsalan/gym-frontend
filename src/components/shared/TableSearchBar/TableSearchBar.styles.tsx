import { Fab, Select, selectClasses, TextField, Toolbar } from '@mui/material';

import { styled } from '@mui/material/styles';

export const StyledTextField = styled(TextField)({
	width: '100%',
	input: {
		padding: '10px 10px',
	},
});

export const StyledSelectField = styled(Select)(() => ({
	width: '100%',
	[`.${selectClasses.select}`]: {
		padding: '10px 10px',
	},
}));

export const StyledToolbar = styled(Toolbar)({
	padding: '20px',
});

export const CustomFab = styled(Fab)({
	position: 'absolute',
	top: '55px',
	right: 0,
});
