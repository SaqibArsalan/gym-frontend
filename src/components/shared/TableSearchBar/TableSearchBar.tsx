import React, { useState } from 'react';

import { Paper, TextFieldProps, Typography } from '@mui/material';

import AddIcon from '@mui/icons-material/Add';
import { IFilterDropdown, IFilterOption, ITableSearchBarProps } from './TableSearchBar.interface';

import {
	StyledTextField,
	StyledToolbar,
	CustomFab,
} from './TableSearchBar.styles';
import './TableSearchBar.scss';
import { applyFilterHOC, InputAdornmentConfiguration } from './TableSearchBar.configs';

const FlteredTextFieldComponent = applyFilterHOC(StyledTextField as React.FC<TextFieldProps>);

const TableSearchBar = (props: ITableSearchBarProps) => {
	const {
		customInputComponent,
		customInputComponentProps,
		showAddButton,
		onAddButtonClick,
		onRemove,
		onChange,
		showRemoveIcon,
		removeIcon,
		searchIcon,
		placeholder,
		defaultValue,
		enableFilter,
		onFilterChange,
		filterOptions,
		defaultSelectedFilterOption
	} = props;

	const [value, setValue] = useState<string>(defaultValue || '');
	const [filterOption, setFilterOption] = useState<IFilterOption | null>(defaultSelectedFilterOption || null);

	const onValueChange = (e: any) => {
		if (e.currentTarget) {
			const text = e.currentTarget.value || '';
			if (onChange) {
				onChange(text, filterOption);
			}
			setValue(text);
		}
	};

	const onRemoveIconClick = () => {
		if (onRemove) {
			onRemove();
		}
		setValue('');
	};

	const onFilterSelection = (filterOption: IFilterOption) => {
		if(onFilterChange) {
			onFilterChange(filterOption);
		}
		if (onChange && value) {
			onChange(value, filterOption);
		}
		setFilterOption(filterOption);
	}

	const inputProps = InputAdornmentConfiguration(
		Boolean(showRemoveIcon && value), 
		onRemoveIconClick, 
		removeIcon, 
		searchIcon
	);
	
	const placeholderValue = enableFilter ? `Search By ${filterOption?.name}` : placeholder;

	const defaultTextFieldProps: TextFieldProps = {
		variant:'outlined',
		id:'search-text',
		placeholder: placeholderValue,
		InputProps: inputProps,
		onChange: onValueChange,
		value,
	};

	const filterProps: IFilterDropdown = {
		options: filterOptions as any,
		onFilterChange: onFilterSelection,
		defaultValue: defaultSelectedFilterOption
	};

	const DefaultFieldWrapper = enableFilter ? 
		FlteredTextFieldComponent
	: StyledTextField;

	const defaultFieldProps = enableFilter ? {
		textFieldProps: defaultTextFieldProps, filterProps
	}: defaultTextFieldProps;

	const ComponentToRender = customInputComponent;

	return (
		<Paper variant='outlined' square>
			<StyledToolbar>
				<div className='toolbar-content-wrapper'>
					<Typography component='div' className='textfield-wrapper'>
						{ComponentToRender ? (
							<ComponentToRender {...customInputComponentProps} />
						) : (
							<DefaultFieldWrapper {...defaultFieldProps as any} />
						)}
					</Typography>
					{showAddButton ? (
						<Typography component='div' className='add-button-wrapper'>
							<CustomFab onClick={onAddButtonClick} color='primary' aria-label='add'>
								<AddIcon />
							</CustomFab>
						</Typography>
					) : null}
				</div>
			</StyledToolbar>
		</Paper>
	);
};

export default TableSearchBar;
