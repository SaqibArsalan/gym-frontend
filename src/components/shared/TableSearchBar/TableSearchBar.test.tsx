import React from 'react';
import "@testing-library/react/dont-cleanup-after-each";
import {cleanup, fireEvent, prettyDOM, queryByAttribute, render, RenderResult } from '@testing-library/react';
import '@testing-library/jest-dom'; 
import TableSearchBar from './TableSearchBar';
import { ITableSearchBarProps } from './TableSearchBar.interface';

let renderWrapper: RenderResult;

const getById = queryByAttribute.bind(null, 'id');

const tableSearchProps: ITableSearchBarProps = {
    onChange: jest.fn(),
    onRemove: jest.fn(),
    showRemoveIcon: true,
    defaultValue: "test"
}

describe("TableSearchBar tests", () => {
    beforeAll(() => {
        renderWrapper = render(<TableSearchBar {...tableSearchProps} />);
    });
    afterAll(() => {
        cleanup();
    });
    test("default renders successfully", () => {
        expect(renderWrapper.container).not.toBeNull();
        expect(renderWrapper.container).toMatchSnapshot();
    });
    test("default render successfully with wrapper and StyledTextField", () => {
        const paperElement = renderWrapper.container.getElementsByClassName('MuiPaper-root').length;
        const toolbarRoot = renderWrapper.container.getElementsByClassName('MuiToolbar-root').length;
        const searchBar = getById(renderWrapper.container, "search-text");
        expect(paperElement).not.toBe(0);
        expect(toolbarRoot).not.toBe(0);
        expect(searchBar).not.toBeNull();
    });

    test("render with onChange trigger", () => {
        const searchBar = getById(renderWrapper.container, "search-text");
        fireEvent.change(searchBar as HTMLElement, {target: {value: 'tests'}});
        expect(tableSearchProps.onChange).toBeCalled();
    });

    test("render with onRemove trigger", () => {
        const removeButton = getById(renderWrapper.container, "remove-button");
        fireEvent.click(removeButton as HTMLElement);
        expect(tableSearchProps.onRemove).toBeCalled();
    });
})