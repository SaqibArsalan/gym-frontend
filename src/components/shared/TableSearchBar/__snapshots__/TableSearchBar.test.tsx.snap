// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableSearchBar tests default renders successfully 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-outlined css-mta8o9-MuiPaper-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-nzn475-MuiToolbar-root"
    >
      <div
        class="toolbar-content-wrapper"
      >
        <div
          class="MuiTypography-root MuiTypography-body1 textfield-wrapper css-ahj2mt-MuiTypography-root"
        >
          <div
            class="MuiFormControl-root MuiTextField-root css-1i2y4ps-MuiFormControl-root-MuiTextField-root"
          >
            <div
              class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedStart MuiInputBase-adornedEnd css-2ehmn7-MuiInputBase-root-MuiOutlinedInput-root"
            >
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionStart MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-ittuaa-MuiInputAdornment-root"
              >
                <span
                  class="notranslate"
                >
                  ​
                </span>
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="SearchIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z"
                  />
                </svg>
              </div>
              <input
                aria-invalid="false"
                class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedStart MuiInputBase-inputAdornedEnd css-152mnda-MuiInputBase-input-MuiOutlinedInput-input"
                id="search-text"
                type="text"
                value="test"
              />
              <div
                class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
              >
                <span
                  aria-label="remove value"
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  id="remove-button"
                  role="button"
                  tabindex="0"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="ClearIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </span>
              </div>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-hdw1oc"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
